File: otUI/CryptDialog.cs
using System;
using Gtk;
using otlib;
using UI = Gtk.Builder.ObjectAttribute;
using otUI;

//[UI] Gtk.TextView tv = new Gtk.TextView();

public class CryptDialog : Gtk.Dialog
{
    Builder builder;
    private AppSettings appSettings;

    [UI] Gtk.TextView keyInput = new Gtk.TextView();
    [UI] Gtk.TextView msgCipherInput = new Gtk.TextView();
    [UI] Gtk.TextView output = new Gtk.TextView();


    public static CryptDialog Create(AppSettings appSettings)
    {
        Builder builder = new Builder(null, "otUI.interfaces.Crypt.ui", null);
        return new CryptDialog(builder, builder.GetObject("cryptdialog").Handle, appSettings);
    }


    protected CryptDialog(Builder builder, IntPtr handle, AppSettings appSettings) : base(handle)
    {
        this.builder = builder;
        this.appSettings = appSettings;

        builder.Autoconnect(this);
        AddButton("Close", ResponseType.Close);
    }

    protected void OnEncryptClicked(object sender, EventArgs e)
    {
        try
        {
            PrettyPrint pp = new PrettyPrint();
            byte[] ks = otp.ToBytes(pp.UnPrettify(keyInput.Buffer.Text), appSettings.CodeCharSetString);
            string msg = msgCipherInput.Buffer.Text;

            byte[] ciphertext = otp.Encrypt(msg, ks, appSettings.CodeCharSetString, appSettings.TextCharSetString);

            if (appSettings.hasPrettyPrint)
            {
                output.Buffer.Text = pp.Prettify(otp.ToString(ciphertext, appSettings.CodeCharSetString));
            }
            else 
            {
                output.Buffer.Text = otp.ToString(ciphertext, appSettings.CodeCharSetString);
            }
        }
        catch (System.Collections.Generic.KeyNotFoundException error)
        {
            ErrorDialog.ShowAlert(this, "Message contains character that is not in textCode characterset");
        }
        catch (Exception error2) 
        {
            ErrorDialog.ShowAlert(this, error2.Message);
        }
    }

    protected void OnDecryptClicked(object sender, EventArgs e)
    {
        PrettyPrint pp = new PrettyPrint();
        byte[] ks = otp.ToBytes(pp.UnPrettify(keyInput.Buffer.Text), appSettings.CodeCharSetString);
        string ciphertext = pp.UnPrettify(msgCipherInput.Buffer.Text);

        byte[] msg = otp.Decrypt(ciphertext, ks, appSettings.CodeCharSetString, appSettings.TextCharSetString);
        output.Buffer.Text = otp.ToString(msg, appSettings.TextCharSetString);
    }

    protected void OnHelpClicked(object sender, EventArgs e)
    {
        HelpDialog hd = HelpDialog.Create(otUI.HelpConst.CryptKeysHelp);
        hd.Run();
        hd.Destroy();
    }

    protected void OnCopyClicked(object sender, EventArgs e)
    {
        Gtk.Clipboard clipboard = Gtk.Clipboard.Get(Gdk.Atom.Intern("CLIPBOARD", false));
        clipboard.Text = output.Buffer.Text;
    }
}

File: UI/CryptDialog.py
import gi

from UI import HelpDialog
from Core.Constants.Help import CRYPT
from Core.PrettyPrint import prettyfy, unprettyfy

gi.require_version('Gdk', '3.0')
gi.require_version('Gtk', '3.0')
gi.require_version('Gio', '2.0')

from gi.overrides import Gdk
from gi.repository import Gtk
from gi.repository import Gio

from Core.Crypto import OTP
from UI.ErrorDialog import ShowAlert
from UI import KeyfileChangePasswordDialog, KeyfileUseKeyDialog
from main import app_settings, resource


Gio.Resource._register(resource)


@Gtk.Template(resource_path="/org/onetimepadui/UI/interfaces/CryptDialog.ui")
class CryptDialog(Gtk.Dialog):
    __gtype_name__ = "CryptDialog"

    output = Gtk.Template.Child()
    key_input = Gtk.Template.Child()
    msg_cipher_input = Gtk.Template.Child()

    def __init__(self):
        super().__init__()
        self.init_template()
        self.otp = OTP.Create(app_settings.get_code_charset(), app_settings.get_text_charset(), app_settings.has_padding)

    def get_text(self, text_view):
        text_buffer = text_view.get_buffer()
        start_iter, end_iter = text_buffer.get_start_iter(), text_buffer.get_end_iter()
        text = text_buffer.get_text(start_iter, end_iter, True)
        return text

    @Gtk.Template.Callback()
    def onEncryptClicked(self, button):
        try:
            msg, key = self.get_text(self.msg_cipher_input), self.get_text(self.key_input)

            cipher_text = self.otp.encrypt(msg, unprettyfy(key))

            if app_settings.has_padding:
                cipher_text = prettyfy(cipher_text)

            self.output.get_buffer().set_text(cipher_text)
        except KeyError:
            ShowAlert(self, "Message contains character that is not in textCode character set")
        except Exception as e:
            ShowAlert(self, str(e))

    @Gtk.Template.Callback()
    def onDecryptClicked(self, button):
        cipher_text, key = self.get_text(self.msg_cipher_input), self.get_text(self.key_input)

        plain_text = self.otp.decrypt(unprettyfy(cipher_text), unprettyfy(key))
        self.output.get_buffer().set_text(plain_text)

    @Gtk.Template.Callback()
    def onCopyClicked(self, button):
        clipboard = Gtk.Clipboard.get(Gdk.SELECTION_CLIPBOARD)
        textbuffer = self.output.get_buffer()
        start_iter, end_iter = textbuffer.get_start_iter(), textbuffer.get_end_iter()
        text = textbuffer.get_text(start_iter, end_iter, True)
        clipboard.set_text(text, -1)

    @Gtk.Template.Callback()
    def onHelpClicked(self, button):
        HelpDialog.main(self, CRYPT)

    @Gtk.Template.Callback()
    def onCloseClicked(self, button):
        self.destroy()

    @Gtk.Template.Callback()
    def onChangePasswordActivate(self, menu_item):
        KeyfileChangePasswordDialog.main()

    @Gtk.Template.Callback()
    def onLoadKeyActivate(self, menu_item):
        key = KeyfileUseKeyDialog.main()
        self.key_input.get_buffer().set_text(key)


def main():
    dialog = CryptDialog()
    dialog.show()

    Gtk.main()
